### PR TITLE
Infer compile time GOPATH from fn.Name().

### PR DESCRIPTION
**Pull request by ChrisHines**
Fix for #8.

All existing tests still pass, and I have tested this against both scenarios mentioned in issue #8 but I cannot think of a convenient way to test the issue directly from within this repository. We could potentially do a bunch of file manipulations to generate a package structure in a temporary folder and run the test as a child process, but that might be going too far. Thoughts?

### CODE REVIEW

**Comment by ChrisHines** [comment]
One side effect of this change is that errors produced in package main will not have their full package path used, but only the last segment of the package path. This is because `runtime.Func.Name` does not include the full import path as a prefix for functions in package main. For example, with this PR vendored, the test repo I posted outputs:

```
pkg-errors-test/pkg-errors-test.go:8: where am I?
```

It would be nice if it produced the following instead.

```
github.com/ChrisHines/pkg-errors-test/pkg-errors-test.go:8: where am I?
```

This result is mildly unfortunate. Errors produced by non-main packages should work as desired. I think it is a net improvement, especially when `github.com/pkg/errors` is vendored.

**Comment by davecheney** [comment]
Thanks Chris, I'm not too worried about this restriction, as it works
towards encouraging people to keep their main packages small.

File: errors.go [hunk 1]
```diff
@@ -65,11 +65,45 @@ func (l loc) Location() (string, int) {
...
+	//    pkg/sub/file.go
+	//
+	// From this we can easily see that fn.Name() has one less path separator
+	// than our desired output. We count separators from the end of the file
+	// path until it finds two more than in the function name and then move
+	// one character forward to preserve the initial path segment without a
+	// leading separator.
+	const sep = "/"
+	goal := strings.Count(fn.Name(), sep) + 2
+	pathCnt := 0
```
Comment (davecheney) [inline errors.go]: Would you move this into the for loop, 

for pathCnt := 0; pathCnt < goal; pathCnt++

And while your there name the loop variable n, not pathCnt

File: errors.go [hunk 2]
```diff
@@ -65,11 +65,45 @@ func (l loc) Location() (string, int) {
...
+	i := len(file)
+	for pathCnt < goal {
+		i = strings.LastIndex(file[:i], sep)
+		if i == -1 {
+			i = -len(sep)
```
Comment (davecheney) [inline errors.go]: I -= len(sep)

**Comment by ChrisHines** [comment]
I am not worried about it either. Just providing full disclosure. This logic has been in github.com/go-stack/stack and github.com/inconshreveable/log15 for a long time and I've never received any complaints.

File: errors.go [hunk 3]
```diff
@@ -65,11 +65,45 @@ func (l loc) Location() (string, int) {
...
+	// We want to produce:
+	//
+	//    pkg/sub/file.go
+	//
+	// From this we can easily see that fn.Name() has one less path separator
+	// than our desired output. We count separators from the end of the file
+	// path until it finds two more than in the function name and then move
+	// one character forward to preserve the initial path segment without a
+	// leading separator.
+	const sep = "/"
```
Comment (davecheney) [inline errors.go]: I cannot test atm, but does this work on windows?

**Comment by davecheney** [comment]
Thanks Chris, this looks pretty good with some very minor nit picking comments

File: errors.go (hunk 3 above)
Comment (ChrisHines) [inline errors.go]: Yes. Oddly package runtime uses "/" for these paths even on Windows. I wrote and tested this on Windows, but have used this logic also on Linux.

Working on fixing the nits. Not sure why that variable wasn't part of the for loop to begin with.

File: errors.go (hunk 2 above)
Comment (ChrisHines) [inline errors.go]: That doesn't do the same thing.

Comment (ChrisHines) [inline errors.go]: I will add a comment to clarify the logic.

**Comment by ChrisHines** [comment]
@davecheney PTAL

File: errors.go (hunk 3 above)
Comment (davecheney) [inline errors.go]: Thanks for checking.

**Comment by davecheney** [comment]
LGTM. Thanks!

**Pull request merged**